File: leetcode/100-999/144. Binary Tree Preorder Traversal/solution_1.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */


/**
 * @param {number[]} tree
 * @param {TreeNode} root
 * @return {void}
 */

function walkTree(tree, root)
{
    if (root)
    {
        tree.push(root.val)

        walkTree(tree, root.left)
        walkTree(tree, root.right)
    }
}

/**
 * @param {TreeNode} root
 * @return {number[]}
 */

var preorderTraversal = function(root) 
{
    const tree = [] 
    walkTree(tree, root)
    return tree
};

File: leetcode/100-999/144. Binary Tree Preorder Traversal/solution_1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution
{
    private List<Integer> tree = new ArrayList<>();
    
    public List<Integer> preorderTraversal(TreeNode root)
    {
        if (root != null)
        {
            tree.add(root.val);
            
            preorderTraversal(root.left);
            preorderTraversal(root.right);
        }
        
        return tree;
    }
}